README.md
# dns-performance-test

This script will query DNS servers with a provided query list and measure the response time.

## Device Identification

In order to track multiple results from different hosts, the hostname and a unique ID is gathered from each host where the script runs.

For the UUID, when the script runs for it's initial attempt, it'll try and see if there's a uuid.cfg file. 
* If there is a uuid.cfg file, it'll read the contents and use that for identification.
* if there is not a uuid.cfg file, it'll create a new uuid and write that to uuid.cfg.

## UUID
To read the current UUID for the device:
```bash
python3 dns-resolution-test.py --getUuid
```

To delete the UUID, ether remove the uuid.cfg file, or run the command:
```bash
python3 dns-resolution-test.py --deleteUuid
```

To renew/reset the UUID:
```bash
python3 dns-resolution-test.py --renewUuid
```

## Tags

If you want to tag your results for aggregating data across multiple devices, do the following:
```bash
python3 dns-resolution-test.py --setTag <TAGNAME>
```

If there's no tag.cfg file, then deviceTag in the output.json file is "".

If you want to remove the tag, either remove the tag.cfg file or:
```bash
python3 dns-resolution-test.py --deleteTag
```

To display the current tag:
```bash
python3 dns-resolution-test.py --getTag
```

## Reference files

By default, it'll read the inputs from filenames:
- nameservers.txt - list of DNS nameservers to query
- queries.txt - list of queries to perform on the DNS nameservers.

**It will also by default not create any output (file, stdout, other).**

The queries file will accept entries in the following formats:
```
test.com
123.123.123.123,ptr
abc.com,a
cnn.com,mx
google.com,aaaa
```

If an entry has an incorrect DNS query type, for example - referencing 'aaa' (invalid) instead of 'aaaa' (valid IPv6 query type), the program will not stop, but will report an error and continue to the next entry.

## Arguments

If you want output there are some options available (in any combination):
```
  --ifname IFNAME      List of nameserver IP addresses file, each entry on a new line. This can be a URL as well.
  --ifquery IFQUERY    List of queries file to be performed, each entry on a new line. This can be a URL as well.
  --ofresults          JSON results output file (uuid,tag,script start time, script end time, results)
  --jsonstdout         print results to stdout
  --displayResponses   Display formatted results
  --verbose            Displays the response times of all the tests.
  --setTag SETTAG      Set the tag for the query results. Creates tag.cfg file with tag.
  --deleteTag          Delete the tag file - tag.cfg
  --getTag             Get the tag from tag.cfg file
  --renewUuid          Renew the UUID value.
  --getUuid            Get the UUID value from uuid.cfg file.
  --deleteUuid         Remove the UUID value. Caution: when script runs again a new UUID will be generated.
  --httpPOST HTTPPOST  Upload the JSON results to the URL
```


## JSON Sample format

```json
{
  "deviceUuid": "<UUID>",
  "hostName": "<HOSTNAME>",
  "deviceTag": "<DEVICETAG>",
  "scriptUTCStartTime": "<Script start time in UTC Format>",
  "scriptUTCEndTime": "<Script end time in UTC Format>",
  "dataFormatVersion": 3,
  "queryResults": {
    "dnsNameServerIP": [
      {
        "query": {
          "queryType": "hostname.example"
        },
        "response": [
          "<IP1>"
        ],
        "responseTime": "<Time in Milliseconds for response>",
        "responseTTL": <time_in_seconds_from_nameserver>
      },
      {
        "query": {
          "queryType": "hostname2.example2"
        },
        "response": [
          "<IP1>",
          "<IP2>"
        ],
        "responseTime": "<Time in Millseconds for response>",
        "responseTTL": <time_in_seconds_from_nameserver>
      },
      {
        "query": {
          "queryType": "hostname3.example3"
        },
        "response": [
          "<IP1>",
          "<IP2>",
          "<IP3>",
          "<IP4>"
        ],
        "responseTime": "<Time in Millseconds for response>",
        "responseTTL": <time_in_seconds_from_nameserver>
      }
    ]
  }
}
```

## Actual Sample Result:

```json
{
  "deviceUuid": "c166d274-2fca-42c2-9a26-d1887ca97bb2",
  "hostName": "myTestHost",
  "deviceTag": "production",
  "scriptStartTime": "2021-05-22 20:25:49.706083",
  "scriptEndTime": "2021-05-22 20:25:49.748855",
  "dataFormatVersion": 3,
  "queryResults": {
    "8.8.8.8": [
      {
        "query": {
          "a": "test.com"
        },
        "response": [
          "69.172.200.235"
        ],
        "responseTime": "7.1",
        "responseTTL": 311
      },
      {
        "query": {
          "a": "google.com"
        },
        "response": [
          "142.250.217.110"
        ],
        "responseTime": "13.3",
        "responseTTL": 3429
      },
      {
        "query": {
          "a": "abc.com"
        },
        "response": [
          "99.84.73.44",
          "99.84.73.60",
          "99.84.73.41",
          "99.84.73.70"
        ],
        "responseTime": "21.8",
        "responseTTL": 151
      },
      {
        "query": {
          "mx": "testdomain.com"
        },
        "response": [
          "20 mail.testdomain.com.",
          "20 mail2.testdomain.com.",
          "30 mail3.testdomain.com."
        ]   
      }
    ]
  }
}
```


Future improvements:
* default list of nameservers and queries to be included in bundle
* ability to publish results to google sheets, ms excel online, etc.
* show if the DNS lookup is a PASS/FAIL based on evaluating criteria.
    * evaluation criteria:
        * ALL responses must be the same
        * ANY of the responses must be what we expect
    * ability to set an evaluation file based on immediate results for comparison later.
    * show current date, and delta compared to when initial evaluation file was created.
* support for DNSSEC queries


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.21 - 2022-12-04
### Changed
* Improved python syntax formatting.
* Improved interpretation of DNS responses.

## 0.20 - 2022-01-23
### Changed
* Fixed display output to be better formatted.

## 0.19 - 2021-08-24
### Changed
* Fixed bug around UUID and Tag handling.
* Improved object names and comments in script.

## 0.18 - 2021-08-22
### Added
* Added the ability to resolve different DNS requests - SOA, NS, CNAME, A, AAAA, PTR, MX. By default an entry in queries file that doesn't have a DNS type will be assumed to be an 'A' record. Refer to readme for format of queries file.
* Added --displayResponses option with column formatting. 
* With newer improvements for DNS queries, included a data structure format version tag in json output.

### Changed
Updated the display results formatting when run from command line.


## 0.17 - 2021-08-15
### Added
Added the ability to download nameservers and queries entries from a webserver (both HTTP and HTTPS). I have not added support for error handling when the hostname in the URL is invalid.

## 0.16 - 2021-08-14
### Added
Allows the ability to upload the json data through the HTTP POST method.

## 0.15 - 2021-08-14
### Fixed
Fixed a minor bug with a_responseTTL variable and a DNS nameserver response timing out.

## 0.14 - 2021-08-14
### Added
Included response time-to-live (TTL) from name server.

## 0.13 - 2021-05-29
### Added
Improved code comments for better programming interpretation.

## 0.12 - 2021-05-22
### Added
Added the ability to manage tags from device.

## 0.11 - 2021-05-22
### Changed
Minor change logic around how arguments are parsed.

## 0.10 - 2021-05-22
### Added
Added options for managing UUID and tag files.


## 0.10 - 2021-05-22
### Changed
Fixed logic for print results to files, stdout, etc.

## 0.09 - 2021-05-22
### Added
* Ability to tag results for aggregating data. To use, create a 'tag.cfg' file in the same working directory as dns-resolution-test.py and insert a label. For example:
```bash
echo "production" > tag.cfg
```
If the tag.cfg file does not exist then the deviceTag field is "".


## 0.08 - 2021-05-22
### Added
* Added script start and end time (UTC format) to json results.
* Added hostname gathering feature to json results
* Added UUID creation feature to track results across multiple devices.

## 0.07 - 2021-05-20

### Added
Added argument option to print results in json format to stdout

### Changed
Amended results field from DNS query to "Err" if there's a:
* DNS Timeout
* No response


## 0.06 - 2021-05-20
### Changed
Minor adjustment to output data to output.json, not output.txt

## 0.05 - 2021-05-01
### Added
Added error handling for dns.resolver.NoNameservers

## 0.04 - 2021-04-30 
### Changed
Improved error handling and display output. Display entries being loaded from files. 

## 0.03 - 2021-04-17
### Added
Added ability to display verbose results.

### Changed
None

### Fixed
None

## 0.02 - 2021-04-16

### Added
Added ability to grab custom input files for nameservers and queries.

### Changed
None

### Fixed
None

## 0.01 - 2021-04-14

Initial build.

### Added
Added ability to perform DNS queries to multiple name servers and measure the response time.


### Changed
None

### Fixed
None
 
 
 


dns-resolution-test.py
# DNS Performance Testing
# Version:            0.21
# Last updated:       2022-12-04

scriptVersion = "0.21"

import sys
import argparse
import json
import dns.resolver
from datetime import datetime, timedelta
import os.path
from os import path

from systemInfo import systemInfo, systemData

import requests

# Global Variables
dnsResponseTextMaxLength = 0

def writeResults(results, outputFile):
    """
    Send the json data to the outputFile variable.
    """
    outputfile = open(outputFile, "w", encoding="utf-8")
    
    outputfile.write(json.dumps(results)) 
    
    outputfile.close()

def printJsonStdout(results):
    """
    This will output the json data to stdout.
    """
    print(json.dumps(results))
    print()

def uploadJsonHTTP(url, sonData):
    """
    This will upload the json data to a URL via a POST method.
    If the verbose argument is set, it'll display what URL it's being 
    submitted to as well as the json data (jsonData).
    When the response is returned, it'll return the X-Headers that are sent back
    from the server.
    """   
    x = requests.post(url, json = jsonData)
    if args.verbose:
        print('Submission URL: ', url)
        print('jsonData: ', json.dumps(jsonData))
        print('X-Headers: ', x.headers)
    return x.headers

def getFileFromURL(fileURL):
    """
    This function will download the contents of fileURL and return a list with the contents.
    """
    tmpData = []
    try:
        urlData = requests.get(fileURL)
        if urlData.status_code == 200:
            tmpData = urlData.text.split('\n')
            try:
                """
                Attempt to remove any blank entries in the dict
                """
                tmpData.remove('')
            except ValueError:
                """
                Do nothing
                """
        else:
            tmpData = ['Error while retrieving URL']
    except socket.gaierror:
        print('Invalid hostname')
        tmpData = ['URL error']
    except requests.exceptions.Timeout:
        print('Timeout while retrieving URL.')
        tmpData = ['URL Timeout']
    except requests.exceptions.TooManyRedirects:
        print('Too many redirects while accessing the URL')
        tmpData = ['URL Redirects too many']
    except requests.exceptions.ConnectionError:
        print('Could not connect to URL - ' + fileURL + '\n')
        tmpData = ['URL connection error']
    return tmpData


def loadNameServersFile(nameserversFile):
    """
    This will load the name servers from the file nameserversFile.
    Each name server should be on it's own line.
    """
    
    dnsNameServers = []
    
    """
    Check to see if nameserversFile is a URL and if it is, attempt to download it
    """

    if nameserversFile.startswith('http://') or nameserversFile.startswith('https://'):
        dnsnameServers = getFileFromURL(nameserversFile)
        return dnsnameServers
   

    if not path.exists(nameserversFile):
        print('I cannot find file ' + nameserversFile)
        sys.exit(1)

    if args.verbose:
        print('Loading the nameservers that are to be queried.')
    
    nameServerFile = open(nameserversFile, "r", encoding="utf-8")

    for line in nameServerFile:
        if args.verbose:
            print(line.rstrip('\n'), end=' ')
        dnsNameServers.append(line.rstrip('\n'))
    if args.verbose:
        print()

    return dnsNameServers


def loadQueriesFile(queriesFile):
    """
    This will load the queries that need to be performed against each name server.
    One query per line. Right now it's only meant to be for 'A' record resolutions.
    """
    queries = []


    """
    Check to see if queriesFile is a URL and if it is, attempt to download it
    """
    if queriesFile.startswith('http://') or queriesFile.startswith('https://'):
        queries = getFileFromURL(queriesFile)
        return queries


    """
    Check to see if if the file exists. If not, exit with error code 1.
    """
    if not path.exists(queriesFile):
        print('I cannot find file ' + queriesFile)
        sys.exit(1)

    if args.verbose:
        print('Loading queries from ' + queriesFile)

    

    queryFile = open(queriesFile, "r", encoding="utf-8")

    for line in queryFile:
        if ","  in line:
            tmpLine = line.rstrip('\n').split(',')
            queries.append({tmpLine[1]: tmpLine[0]})
        else:
            queries.append({'a': line.rstrip('\n')})
        
        if args.verbose:
            print(line.rstrip('\n'), end=' ')
        
    
    if args.verbose:
        print()
        print(queries)

    queryFile.close()


    return queries



def displayResults(results):
    """
    This will display the results to stdout. Not always formatted correctly 
    because the responses could have a variable number.
    """

    # Get the global variable dnsResponseTextMaxLength
    global dnsResponseTextMaxLength
    filler = ' '

    if args.verbose:
        print("dnsResponseTextMaxLength = ", dnsResponseTextMaxLength)

    headers = ['DNS Server', 'DNS Type', 'DNS Query', 'DNS Response', 'Response Time (ms)', 'DNS TTL']

    """
    DNS Column lengths
    """
    dnsServerLength = 18
    dnsQueryTypeLength = 15
    dnsQueryLength = 30
    dnsResponseTextMaxLength += 10
    dnsResponseTimeLength = 20
    dnsResponseTTLLength = 8



    """
    Print the headers with appropriate spacing.
    """

    for count, tem in enumerate(headers):
        if count == 0:
            # DNS Server
            print(f'{item:{filler}<{dnsServerLength}}', end='')

        if count == 1: 
            # DNS Type
            print(f'{item:{filler}<{dnsQueryTypeLength}}', end='')

        if count == 2:
            # DNS Query
            print(f'{item:{filler}<{dnsQueryLength}}', end='')

        if count == 3:
            # DNS Response
            print(f'{item:{filler}<{dnsResponseTextMaxLength}}', end='')

        if count == 4:
            # Response Time (ms)
            print(f'{item:{filler}<{dnsResponseTimeLength}}', end='')

        if count == 5:
            # DNS TTL
            print(f'{item:{filler}<{dnsResponseTTLLength}}', end='')


    print()

    """
    Iterate through the response data and format the columns with appropriate spaces.
    """
    for nameserverItem in results:

        for dataItem in results[nameserverItem]:
            print(f'{nameserverItem:{filler}<{dnsServerLength}}', end='')

            for dataItem2 in dataItem:
                #print(str(dataItem[dataItem2]))
                if dataItem2 == 'query':
                    queryType = list(dataItem[dataItem2].items())[0][0]
                    queryName = list(dataItem[dataItem2].items())[0][1]
                    print(f'{queryType:{filler}<{dnsQueryTypeLength}}', end='')
                    print(f'{queryName:{filler}<{dnsQueryLength}}', end='')

                if dataItem2 == 'response':
                    dnsElements = list(dataItem[dataItem2])

                    dnsResponseFormatted = ','.join(dnsElements)
                    print(f'{dnsResponseFormatted:{filler}<{dnsResponseTextMaxLength}}', end='')

                if dataItem2 == 'responseTime':
                    responseTime = dataItem[dataItem2]
                    print(f'{responseTime:{filler}<{dnsResponseTimeLength}}', end='')

                if dataItem2 == 'responseTTL':
                    responseTTL = dataItem[dataItem2]
                    print(f'{responseTTL:{filler}<{dnsResponseTTLLength}}')



def performQueries(nameservers, queries):
    """
    This will perform all the all the queries against each nameserver.
    """

    # Set the resolver
    resolver = dns.resolver.Resolver()
    
    # Set the results to empty dict
    results = {}

    counter = 1

    totalQueries = len(nameservers) * len(queries)

    # Get the global variable dnsResponseTextMaxLength
    global dnsResponseTextMaxLength

    for server in nameservers:
        # Set the name servers to a single list entry.
        # We don't need multiple failover nameservers because we want 
        # a result from each name server and want to point out any failures.
        resolver.nameservers = [server]
        for query in queries:
            # Display the query count to keep track of progress.
            queryType = list(query.items())[0][0].lower()
            queryName = list(query.items())[0][1].lower()

            if args.verbose:
                print('Query = ' + str(query))
                print('Query count = ' + str(counter) + ' of ' + str(totalQueries))

            # Start Query Time
            queryStartTime = datetime.now()

            try:
                if queryType == "ptr":
                    answer = resolver.resolve_address(queryName)
                else:
                    answer = resolver.resolve(queryName, queryType)
            
            # If there's a timeout, display the timeout, which query and which nameserver
            # typical timeout is 5.5s

            except dns.rdatatype.UnknownRdatatype:
                print('Unkown DNS response - ' + str(query) + ' @' + server)
                answer = []

            except dns.resolver.NoAnswer:
                # No answer from DNS server
                print('No DNS answer - ' + str(query) + ' @' + server)
                answer = []

            except dns.exception.Timeout:
                # DNS Server timed out
                print('DNS Timeout - ' + str(query) + ' @' + server)
                answer = []

            except dns.resolver.NXDOMAIN:
                # NXDOMAIN response from DNS server
                print('NXDOMAIN response - ' + str(query) + ' @' + server)
                answer = []

            except dns.resolver.NoNameservers:
                # If there are no responses from name servers, display No Response.
                print('No response. ' + str(query) + ' @' + server)
                answer = []

            # End query time. 
            queryEndTime = datetime.now()

            # Calculate the difference between End Query Time and Start Query Time
            # Multiply the result by a 1000 to get millisecond response.
            queryTime = (queryEndTime - queryStartTime).total_seconds() * 1000

            # Format the queryTime response.
            s_queryTime = str("{:.1f}".format(queryTime))
            
            l_response = []
           
            # If there were any answers from the query, append the response 
            # to l_response.
            # If there were any errors, add 'Err'
            if answer:
                if queryType == "a" or queryType == "aaaa":
                    for counter, response in enumerate(answer):
                        aResponse = response.address
                        l_response.append(aResponse)
                        currentLength = len(str(aResponse)) * (counter + 1)
                        if dnsResponseTextMaxLength < currentLength:
                            dnsResponseTextMaxLength = currentLength

                if queryType == "mx":
                    for response in answer:
                        mxResponse = response.to_text()
                        l_response.append(mxResponse)
                        currentLength = len(str(mxResponse))
                        if dnsResponseTextMaxLength < currentLength:
                            dnsResponseTextMaxLength = currentLength

                if queryType == "ptr":
                    for response in answer:
                        ptrResponse = response.to_text()
                        l_response.append(ptrResponse)
                        currentLength = len(str(ptrResponse))
                        if dnsResponseTextMaxLength < currentLength:
                            dnsResponseTextMaxLength = currentLength
                            
                if queryType == "soa":
                    for response in answer:
                        soaResponse = response.to_text().split(' ')[0]
                        l_response.append(soaResponse)
                        currentLength = len(str(soaResponse))
                        if dnsResponseTextMaxLength < currentLength:
                            dnsResponseTextMaxLength = currentLength

                if queryType == "cname":
                    for response in answer:
                        cnameResponse = response.to_text()
                        l_response.append(cnameResponse)
                        currentLength = len(str(cnameResponse))
                        if dnsResponseTextMaxLength < currentLength:
                            dnsResponseTextMaxLength = currentLength

                if queryType == "ns":
                    for response in answer:
                        nsResponse = response.to_text()
                        l_response.append(nsResponse)
                        currentLength = len(str(nsResponse))
                        if dnsResponseTextMaxLength < currentLength:
                            dnsResponseTextMaxLength = currentLength

                a_responseTTL = answer.rrset.ttl
            else:
                a_responseTTL = -1
                l_response.append('Err')
                if dnsResponseTextMaxLength < 3:
                    dnsResponseTextMaxLength = 4

            # If the server is not in the results, set the response to a blank list.
            if server not in results:
                results[server] = []

            try:
                # Create the json dict with all of the responses.
                thisQuery = {"query": query, "response": l_response, "responseTime": s_queryTime, "responseTTL": a_responseTTL}
                results[server].append(thisQuery)
                
                counter += 1

            except KeyError:
                # If there's some accidental programming error, then display error.
                print('I made an oops.')

    return results

def gatherData(queryResults, scriptStartTime, scriptEndTime):
    """
    This will collect all the data into a uniform data structure that can 
    help with measuring results across multiple executions. 

    Data that is included is:
    * deviceUuid         - a unique device identifier.
    * deviceTag          - a tag for the device to help with aggregating data across.
                           multiple endpoints (for example all production, development, qa devices).
    * hostName           - the hostname of the device where script is executing.
    * scriptUTCStartTime - script start time (UTC format).
    * scriptUTCEndTime   - script end time (UTC format).
    * queryResults       - The results of all queries that were performed against the nameservers.  
    """
    myInfo = systemInfo.systemInfo()
    
    if myInfo.uuid == "":
        n = systemData.systemData()
        n.createUuidIfNotExist()
        myInfo.uuid = myInfo.getUuid()

    myData = {
        "deviceUuid": myInfo.uuid,
        "deviceTag": myInfo.deviceTag,
        "hostName": myInfo.hostname,
        "scriptUTCStartTime": scriptStartTime,
        "scriptUTCEndTime": scriptEndTime,
        "dataFormatVersion": 3,
        "queryResults": queryResults
    }

    return myData

def parseArguments():
    """
    Create argument options and parse through them to determine what to do with script.
    """

    # Instantiate the parser
    global scriptVersion
    parser = argparse.ArgumentParser(description='DNS Performance Testing ' + scriptVersion)

    # Optional arguments
    parser.add_argument('--ifname', default="nameservers.txt",
                        help='List of nameserver IP addresses, each on a new line')

    parser.add_argument('--ifquery', default='queries.txt',
                        help='List of queries to be performed.')

    parser.add_argument('--ofresults', action='store_true',
                        help='JSON results output file')

    parser.add_argument('--jsonstdout', action='store_true',
                        help='print results to stdout')

    parser.add_argument('--displayResponses', action='store_true',
                        help='Display formatted results')

    parser.add_argument('--verbose', action='store_true',
                        help='Displays the response times of all the tests.')

    parser.add_argument('--setTag', default='',
                        help='Set the tag for the query results. Creates tag.cfg file with tag.')

    parser.add_argument('--deleteTag', action='store_true',
                        help='Delete the tag file - tag.cfg')

    parser.add_argument('--getTag', action='store_true',
                        help='Get the tag from tag.cfg file')

    parser.add_argument('--renewUuid', action='store_true',
                        help='Renew the UUID value.')

    parser.add_argument('--getUuid', action='store_true',
                        help='Get the UUID value from uuid.cfg file.')

    parser.add_argument('--deleteUuid', action='store_true',
                        help='Remove the UUID value. Caution: when script runs again a new UUID will be generated.')

    parser.add_argument('--httpPOST', default='',
                        help='Upload the JSON results to the URL')

    global args
    args = parser.parse_args()


def defineInfoArguments(o_systemData, o_systemInfo):
    global args
    # If setTag argument is set, create the new Tag.
    if args.setTag:
        o_systemData.setTag(args.setTag)
        print('New tag set.')
        sys.exit(0)

    # If getTag is set, it will grab the value in tag.cfg file.
    if args.getTag:
        print(o_systemInfo.getTag())
        sys.exit(0)

    # If deleteTag is set, it will delete the tag.cfg file.
    if args.deleteTag:
        o_systemData.deleteTag()
        sys.exit(0)

    # If getUuid is set, it grab the value in uuid.cfg
    if args.getUuid:
        print(o_systemInfo.getUuid())
        sys.exit(0)

    # If deleteUuid is set, the uuid.cfg file will be deleted.
    if args.deleteUuid:
        o_systemData.deleteUuid()
        sys.exit(0)

    # If renewUuid is set, first delete uuid.cfg file, then generate a new uuid.
    if args.renewUuid:
        o_systemData.deleteUuid()
        o_systemData.createUuidIfNotExist()
        sys.exit(0)


def main():
    # Parse all the arguments
    parseArguments()
    
    # Gather the information from the device where the script is executed.
    o_mySystemData = systemData.systemData()
    o_myInfo = systemInfo.systemInfo()

    defineInfoArguments(o_mySystemData, o_myInfo)


    # Script start time (UTC format)
    scriptStartTime = datetime.utcnow()

    # If verbose argument is set, display the script start time to stdout.
    if args.verbose:
        print('Script start time: ', str(scriptStartTime), '\n')
        
    
    # Query file from ifquery argument
    queryFile = args.ifquery

    # Nameserver file from ifname argument
    nameserversFile = args.ifname

    # Results to be written to output.json
    outputFileResults = 'output.json'
    
    queries = loadQueriesFile(queryFile)
    nameservers = loadNameServersFile(nameserversFile)
    
    results = performQueries(nameservers, queries)

    # If verbose argument is parsed, display the results to stdout.
    if args.verbose:
        print(results)
    
    if args.displayResponses:
        displayResults(results)

    # Script end time (UTC format)
    scriptEndTime = datetime.utcnow()
    
    # If verbose argument is set, display script end time.
    if args.verbose:
        print('\nScript stop time: ', str(scriptEndTime))
    
    # Collate all the data into myData
    myData = gatherData(results, str(scriptStartTime), str(scriptEndTime))
    
    # If the httpPOST argument is set, send the json data to the URL via POST method
    if args.httpPOST:
        print(uploadJsonHTTP(args.httpPOST, myData))

    # If the jsonstdout argument is set, then print myData to stdout.
    if args.jsonstdout:
        printJsonStdout(myData)
    
    # If the ofresults is set, output the data to outputFileResults.
    # Results will always be overwritten.
    if args.ofresults:
        writeResults(myData, outputFileResults)

if __name__ == '__main__':
    try:
        main()

    except KeyboardInterrupt:
        print('Interrupted')
        print
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)


systemInfo/systemInfo.py
import socket
import uuid
import os.path
from os import path

class systemInfo:

    uuidFilename = 'uuid.cfg'
    tagFilename = 'tag.cfg'

    def generateUuid(self):
        return str(uuid.uuid4())

    def createUuidIfNotExist(self):
        if not path.exists(self.uuidFilename):
            newUuid = self.generateUuid()
            f_uuid = open(self.uuidFilename,'w')
            f_uuid.write(str(newUuid))
            f_uuid.close()


    def getUuid(self):
        self.createUuidIfNotExist()
        f_uuid = open(self.uuidFilename,'r')
        myUuid = f_uuid.readline()
        f_uuid.close()
        return myUuid

    def getTag(self):
        if path.exists(self.tagFilename):
            f_tag = open(self.tagFilename,'r')
            myTag = f_tag.readline().rstrip()
            f_tag.close()
            return myTag
        else:
            return ""

    def __init__(self):
        self.hostname = socket.gethostname()
        self.uuid = self.getUuid()
        self.deviceTag = self.getTag()
    


json2excel/json2excel.py
# JSON to Excel spreadsheet
# Version:            0.02
# Last updated:       2022-12-04

import xlsxwriter
import json
import os.path
from os import path
import sys
import argparse

def copyJsonFile2Excel(jsonFile, excelFile):
    print('Starting conversion...')
    if path.exists(jsonFile):
        jf = open(jsonFile, 'r', encoding='utf-8')
    else:
        print('json file does not exist!')
        sys.exit(1)

    workbook = xlsxwriter.Workbook(excelFile)

    worksheet = workbook.add_worksheet()

    row = 0
    col = 0

    # Create the format options for the workbook.
    headersFormat = workbook.add_format({'bold': 1})
    responseTimeFormat = workbook.add_format({'bold': 1, 'font_color': 'red'})
    numberFormat = workbook.add_format({'num_format': '0.0'})

    # Display initial information
    worksheet.write(row, col, 'Input file:')
    worksheet.write(row, col+1, jsonFile)
    worksheet.write(row+1, col, 'Output file:')
    worksheet.write(row+1, col+1, excelFile)
    worksheet.write(row+2, col, 'Number of Nameserver Entries:')
    worksheet.write(row+3, col, 'Average response time:')    
    entries =  0

    row = 6

    # Create headers for all data.
    worksheet.write(row, col, 'deviceUuid', headersFormat)
    worksheet.write(row, col+1, 'hostname', headersFormat)
    worksheet.write(row, col+2, 'deviceTag', headersFormat)
    worksheet.write(row, col+3, 'scriptStartTime', headersFormat)
    worksheet.write(row, col+4, 'scriptEndTime', headersFormat)
    worksheet.write(row, col+5, 'Nameserver', headersFormat)
    worksheet.write(row, col+6, 'Query', headersFormat)
    worksheet.write(row, col+7, 'Record', headersFormat)
    worksheet.write(row, col+8, 'Response', headersFormat)
    worksheet.write(row, col+9, 'Response Time', headersFormat)

    # Grab the first line from the file
    jf_line = jf.readline()
    
    # Loop through each line that is read
    while jf_line:
        json_dict  = json.loads(jf_line)

        for nsEntry in json_dict['queryResults']:
            for query in json_dict['queryResults'][nsEntry]:
                queryElements = list(query['query'].items())
                row += 1
                worksheet.write(row, col, json_dict['deviceUuid'])
                worksheet.write(row, col+1, json_dict['hostName'])
                worksheet.write(row, col+2, json_dict['deviceTag'])
                worksheet.write(row, col+3, json_dict['scriptUTCStartTime'])
                worksheet.write(row, col+4, json_dict['scriptUTCEndTime'])
                worksheet.write(row, col+5, nsEntry)
                worksheet.write(row, col+6, queryElements[0][0])
                worksheet.write(row, col+7, queryElements[0][1])
                worksheet.write(row, col+8, ','.join(query['response']))
                worksheet.write_number(row, col+9, float(query['responseTime']), numberFormat)

        # Keep track with the number of entries.
        entries += 1
        
        # Read the next line in the file.
        jf_line = jf.readline()
        
        row += 1
    
    # Close the jsonFile
    jf.close()

    worksheet.write(2, 1, entries)
    responseTimeArea = 'J8:J' + str(row)

    # Create a conditional format to highlight any entries above the average response time.
    worksheet.conditional_format(responseTimeArea, { 'type': 'average', 'criteria': 'above', 'format': responseTimeFormat } )

    # Display the average time at the information section of spreadsheet for comparison purposes.
    worksheet.write(3, 1, '=AVERAGE(' + responseTimeArea + ')', numberFormat)

    # Close the excel spreadsheet.
    workbook.close()
    
    print('Done.')

def parseArguments():
    # Instantiate the parser
    parser = argparse.ArgumentParser(description='json to excel converter')

    # Optional arguments
    parser.add_argument('--jsonfile', default='output.json',
                        help='json formatted results from dns-resolution-test script')

    parser.add_argument('--excelfile', default='output.xlsx',
                        help='the excel file where you want to format the results')

    global args
    args = parser.parse_args()


def main():
    parseArguments()

    jsonFile = args.jsonfile
    excelFile = args.excelfile
    print('Input file argument : ', jsonFile)
    print('Output file argument: ', excelFile)
    
    copyJsonFile2Excel(jsonFile, excelFile)    
    
if __name__ == '__main__':
    try:
        main()
    
    except KeyboardInterrupt:
        print('Interrupted')
        print
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)
